crypto/evp/evp_enc.c

```diff
@@ -444,22 +456,81 @@ int EVP_CipherInit_ex2(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
                        const unsigned char *key, const unsigned char *iv,
                        int enc, const OSSL_PARAM params[])
 {
-    return evp_cipher_init_internal(ctx, cipher, NULL, key, iv, enc, params);
+    return evp_cipher_init_internal(ctx, cipher, NULL, key, iv, enc, 0, params);
 }
 
 int EVP_CipherInit(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
                    const unsigned char *key, const unsigned char *iv, int enc)
 {
     if (cipher != NULL)
         EVP_CIPHER_CTX_reset(ctx);
-    return evp_cipher_init_internal(ctx, cipher, NULL, key, iv, enc, NULL);
+    return evp_cipher_init_internal(ctx, cipher, NULL, key, iv, enc, 0, NULL);
 }
 
 int EVP_CipherInit_ex(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
                       ENGINE *impl, const unsigned char *key,
                       const unsigned char *iv, int enc)
 {
-    return evp_cipher_init_internal(ctx, cipher, impl, key, iv, enc, NULL);
+    return evp_cipher_init_internal(ctx, cipher, impl, key, iv, enc, 0, NULL);
+}
+
+static int evp_cipher_pipeline_init_internal(EVP_CIPHER_CTX *ctx,
+                                             const EVP_CIPHER *cipher,
+                                             const unsigned char *key,
+                                             size_t numpipes,
+                                             const unsigned char **iv, int enc,
+                                             const OSSL_PARAM params[])
+{
+    if (numpipes > EVP_MAX_PIPES) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_TOO_MANY_PIPES);
+        return 0;
+    }
+
+    ctx->numpipes = numpipes;
+
+    if (!evp_cipher_init_internal(ctx, cipher, NULL, NULL, NULL, enc, 1,
+                                  params))
+        return 0;
+
+    if (enc) {
```

Comment:
`EVP_CipherInit` supports `-1` for enc meaning "leave the value unchanged (the actual value of 'enc' being supplied in a previous call)". We lose that ability for `EVP_CipherPipelineInit`. Is that intentional? Should we retain it?

Reply:
Changing this line as
```
    /* ctx->encrypt is set by evp_cipher_init_internal */
    if (ctx->encrypt) {
```